src/coreclr/nativeaot/Runtime/interoplibinterface_objc.cpp

```diff
@@ -0,0 +1,187 @@
+            pThread->SetSuppressGcStress();
+
+        bool result = pTryGetTaggedMemoryCallback(obj, tagged);
+
+        // Revert GC stress mode if we changed it.
+        if (!fGcStressWasSuppressed)
+            pThread->ClearSuppressGcStress();
+
+        pThread->ClearDoNotTriggerGc();
+
+        return result;
+    }
+
+    template <typename T>
+    T TryGetCallbackViaClasslib(_In_ Object * object, _In_ ClasslibFunctionId id)
+    {
+        void* fn = object->get_EEType()
+                        ->GetTypeManagerPtr()
+                        ->AsTypeManager()
+                        ->GetClasslibFunction(id);
+
+        if (fn != nullptr)
+        {
+            fn = ((ObjCMarshalNative::TryGetCallback)fn)();
+        }
+
+        return (T)fn;
+    }
+}
+
+// One time startup initialization.
+bool ObjCMarshalNative::Initialize()
+{
+    s_sLock.Init(CrstObjectiveCMarshalCallouts, CRST_DEFAULT);
+
+    return true;
+}
+
+bool ObjCMarshalNative::RegisterBeginEndCallback(void * callback)
+{
+    _ASSERTE(callback != nullptr);
+
+     // We must be in Cooperative mode since we are setting callbacks that
+     // will be used during a GC and we want to ensure a GC isn't occurring
+     // while they are being set.
+     _ASSERTE(ThreadStore::GetCurrentThread()->IsCurrentThreadInCooperativeMode());
+    
+    GcBeginEndCallbacks * pCallback = new (nothrow) GcBeginEndCallbacks();
+    if (pCallback == NULL)
+        return false;
+    
+    pCallback->m_pCallbackFunction = (ObjCMarshalNative::BeginEndCallback)callback;
+
+    CrstHolder lh(&s_sLock);
+
+    pCallback->m_pNext = m_pBeginEndCallbacks;
+    m_pBeginEndCallbacks = pCallback;
+
+    return true;
+}
+
+bool ObjCMarshalNative::IsTrackedReference(_In_ Object * object, _Out_ bool* isReferenced)
+{
+    *isReferenced = false;
+
+    if (!object->GetGCSafeMethodTable()->IsTrackedReferenceWithFinalizer())
+        return false;
+    
+    auto pIsReferencedCallbackCallback = TryGetCallbackViaClasslib<ObjCMarshalNative::IsReferencedCallback>(
+        object, ClasslibFunctionId::ObjectiveCMarshalGetIsTrackedReferenceCallback);
+
+    if (pIsReferencedCallbackCallback == nullptr)
+        return false;
+
+    void* taggedMemory;
+    if (!TryGetTaggedMemory(object, &taggedMemory))
+        return false;
+    
+    int result = pIsReferencedCallbackCallback(taggedMemory);
+
+    *isReferenced = (result != 0);
+    return true;
+}
+
+void ObjCMarshalNative::BeforeRefCountedHandleCallbacks()
+{
+    CallBeginEndCallbacks();
+}
+
+void ObjCMarshalNative::AfterRefCountedHandleCallbacks()
+{
+    CallBeginEndCallbacks();
+}
+
+void ObjCMarshalNative::OnEnteredFinalizerQueue(_In_ Object * object)
+{
+    auto pOnEnteredFinalizerQueueCallback = TryGetCallbackViaClasslib<ObjCMarshalNative::EnteredFinalizationCallback>(
+        object, ClasslibFunctionId::ObjectiveCMarshalGetOnEnteredFinalizerQueueCallback);
+
+    if (pOnEnteredFinalizerQueueCallback == nullptr)
+        return;
```

Comment:
Same as with above. If we got here and this is null, then we are hiding a bigger logical flaw.

Reply:
This will now assert in `GetCallbackViaClasslibCallback` if the callback is missing.